pyftpdlib/handlers.py

```diff
@@ -3453,10 +3460,24 @@ def __init__(self, conn, server, ioloop=None):
             self._pbsz = False
             self._prot = False
             self.ssl_context = self.get_ssl_context()
+            if self.client_certfile is not None:
+                self.ssl_context.set_verify(VERIFY_PEER |
+                                            VERIFY_FAIL_IF_NO_PEER_CERT |
+                                            VERIFY_CLIENT_ONCE,
+                                            self.verify_certs_callback)
 
         def __repr__(self):
             return FTPHandler.__repr__(self)
 
+        # Cannot be @classmethod, need instance to log
+        def verify_certs_callback(self, connection, x509,
+                                  errnum, errdepth, ok):
+            if not ok:
+                self.log("Bad client certificate detected.")
+            else:
+                self.log("Client certificate is valid.")
+            return ok
```

Comment:
It's true that `handle_failed_ssl_handshake` will be called when client certificate is not provided. Meanwhile, `self.log("Bad client certificate detected.")` will still be called if client does provide a certificate, but it could not pass the check with server one.